src/node/blockstorage.cpp

```diff
@@ -406,6 +406,7 @@ const CBlockIndex* BlockManager::GetFirstStoredBlock(const CBlockIndex& start_bl
 {
     AssertLockHeld(::cs_main);
     const CBlockIndex* last_block = &start_block;
+    assert(last_block->nStatus & BLOCK_HAVE_DATA);
```

Comment:
> So I think if we now require that "The caller must ensure that upper_block has block data available." we should adjust the existing callers to do that.

Is there something you would like to see specifically, like adding asserts at the call site? I thought most of the call sites were just passing the chain tip, so it's hard to see how this could cause a problem unless the tip was pruned, in which case the current assert failure would be seem pretty appropriate

Reply:
I don't think an assert failure is appropiate if having a pruned tip is a valid state that bitcoin core can be in, which seems to be the case, at least in the examples above. For example, it should be possible to call `getblockchaininfo` directly after loading a snapshot from disk, without triggering an assert.

Reply:
Based on the `AssumeUTXO` line. It seems to be expected to not have block data available right after importing the chainstate. And that shouldn't crash the software if the user calls `pruneblockchain` or `getblockchaininfo`.

Reply:
I see, so maybe current code is buggy and returns bad rpc results then? I guess simplest thing to do is just document GetFirstStoredBlock's odd behavior in this case and keep it unchanged, rather than returning nullptr which could segfault, or adding an assertion.

The new GetFirstStoredBlock function should be able to straightforwardly return false in this case, though.

Reply:
> I see, so maybe current code is buggy and returns bad rpc results then? 

yeah.

> I guess simplest thing to do is just document GetFirstStoredBlock's odd behavior in this case and keep it unchanged, rather than returning nullptr which could segfault, or adding an assertion.

Have struggled a bit with this.
I tried to maintain the previous `GetFirstStoredBlock` behavior but it conflicts so much with the function's description: "Find the first stored ancestor of 'start_block' immediately after the last pruned ancestor" that I couldn't come up with a suitable explanation for this "oddity".

Have preferred to transfer the responsibility of checking the 'start_block' data availability to the few callers who use it (only two places), without altering their behavior in any way.

I think that the changes are minimal and help to reduce confusion. But let me know what you guys think.